### PR TITLE
handle params array in the argument list

### LINKED ISSUE #4842: CA2016: Forward cancellation token codefix is incorrect
See https://github.com/dotnet/efcore/issues/22667

cc: @carlossanlop 

given a type that contains overloads like this:
```csharp

public class C{
  public ValueTask<object?> FindAsync(params object?[]? keyValues) {}
  public ValueTask<object?> FindAsync(object?[]? keyValues, CancellationToken cancellationToken)
}
```

and is called like this:
```csharp
async Task M(string[] args, CancellationToken token)
{
    var c = new C();
    var result = await c.FindAsync(5);
}
```

## Actual

today CA2016 suggests this as a codefix:
```csharp
async Task M(string[] args, CancellationToken token)
{
    var c = new C();
    var result = await c.FindAsync(5, cancellationToken: token); // Error CS1503 Argument 1: cannot convert from 'int' to 'object?[]?'
} 
```

## Expected

if items at the previous call site was part of a `params` array then the codefix should wrap those elements inside a `new object[]` expression

```csharp
async Task M(string[] args, CancellationToken token)
{
    var c = new C();
    var result = await c.FindAsync(new object?[] { 5 }, token);
}
```

an alternative could be 
```csharp
async Task M(string[] args, CancellationToken token)
{
    var c = new C();
    var result = await c.FindAsync(new [] { (object?)5 }, token);
}
```

### PR DESCRIPTION
**Pull request by jmarolf**
fixes #4842

TODO:
- [x] figure out why the SyntaxGenerator for VB generates an invalide tree.,

### CODE REVIEW

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.Visitor.cs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                {
+                    syntax = IdentifierName(symbol.SpecialType == SpecialType.System_IntPtr ? "nint" : "nuint");
+                    return true;
+                }
+
+                syntax = null;
+                return false;
+            }
+
+            private static SyntaxToken ToIdentifierToken(string identifier, bool isQueryContext = false)
```
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.Visitor.cs:195]: This is always `false`. Is that intended?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs [hunk 2]
```diff
@@ -2329,6 +2329,48 @@ static async Task M1Async(string s, CancellationToken cancellationToken)
             return VerifyCS.VerifyCodeFixAsync(originalCode, fixedCode);
         }
 
+        [Fact]
+        [WorkItem(4842, "https://github.com/dotnet/roslyn-analyzers/issues/4842")]
+        public Task CS_ParamsArray()
+        {
+            // Local static functions are available in C# >= 8.0
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocationsTests.cs:2336]: Copy/paste error?

Looks like some existing tests in `main` has this unrelated comment as well :smile:

```suggestion
```

**Review by Youssef1313 — CHANGES_REQUESTED** [review]
Can you also add a test for an API shaped like this:

https://docs.microsoft.com/en-us/dotnet/api/system.data.entity.dbset.findasync?view=entity-framework-6.2.0

```csharp
public virtual System.Threading.Tasks.Task<object> FindAsync (params object[] keyValues);
public virtual System.Threading.Tasks.Task<object> FindAsync (System.Threading.CancellationToken cancellationToken, params object[] keyValues);
```

**Comment by jmarolf** [comment]
> Can you also add a test for an API shaped like this:
> 
> https://docs.microsoft.com/en-us/dotnet/api/system.data.entity.dbset.findasync?view=entity-framework-6.2.0
> 
> ```cs
> public virtual System.Threading.Tasks.Task<object> FindAsync (params object[] keyValues);
> public virtual System.Threading.Tasks.Task<object> FindAsync (System.Threading.CancellationToken cancellationToken, params object[] keyValues);
> ```

The current analyzer/codefixer was intentionally designed to not consider this. I'll get this PR in and then follow up with a re-write allowing cancellation tokens to be in any position.

**Comment by mavasani** [comment]
@carlossanlop Can you please review this PR?

**Comment by mavasani** [comment]
Tagging @ryzngard for review

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+                if (!identifier.StartsWith("@", StringComparison.Ordinal))
+                {
+                    token = token.WithAdditionalAnnotations(Simplifier.Annotation);
+                }
+
+                return token;
+            }
+
+            private static string EscapeIdentifier(string identifier)
```
Comment (ryzngard) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs:219]: Should this go in `SyntaxFacts`?

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+                {
+                    syntax = IdentifierName(symbol.SpecialType == SpecialType.System_IntPtr ? "nint" : "nuint");
+                    return true;
+                }
+
+                syntax = null;
+                return false;
+            }
+
+            private static SyntaxToken ToIdentifierToken(string identifier)
```
Comment (ryzngard) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs:195]: Should this go in `SyntaxFacts`?

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+                    else
+                    {
+                        var container = symbol.ContainingNamespace.Accept(this)!;
+                        typeSyntax = AddInformationTo(QualifiedName(
+                            (NameSyntax)container,
+                            simpleNameSyntax));
+                    }
+                }
+
+                if (symbol.NullableAnnotation() == NullableAnnotation.Annotated &&
```
Comment (ryzngard) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs:82]: Is it possible for `symbol.NullableAnnotation()` to return `Annotated` for value types?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Fixer.cs [hunk 6]
```diff
@@ -111,10 +119,25 @@ Task<Document> CreateChangedDocumentAsync(CancellationToken _)
...
-            ImmutableArray<SyntaxNode> newArguments)
+            ImmutableArray<TArgumentSyntax> currentArguments,
+            IArrayTypeSymbol? paramsArrayType)
         {
             SyntaxGenerator generator = SyntaxGenerator.GetGenerator(doc);
 
+            ImmutableArray<SyntaxNode> newArguments;
+            if (paramsArrayType is not null)
+            {
+                // current callsite is a params array, we need to wrap all these arguments to preserve sematnics
```
Comment (ryzngard) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Fixer.cs:130]:
```suggestion
                // current callsite is a params array, we need to wrap all these arguments to preserve semantics
```

File: src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.vb [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+                Dim elementTypeSyntax = underlyingNonArrayType.Accept(Me)
+                Dim ranks = New List(Of ArrayRankSpecifierSyntax)()
+                Dim arrayType = symbol
+                While arrayType IsNot Nothing
+                    Dim commaCount = Math.Max(0, arrayType.Rank - 1)
+                    Dim commas = SyntaxFactory.TokenList(Enumerable.Repeat(SyntaxFactory.Token(SyntaxKind.CommaToken), commaCount))
+                    ranks.Add(SyntaxFactory.ArrayRankSpecifier(SyntaxFactory.Token(SyntaxKind.OpenParenToken), commas, SyntaxFactory.Token(SyntaxKind.CloseParenToken)))
+                    arrayType = TryCast(arrayType.ElementType, IArrayTypeSymbol)
+                End While
```
Comment (ryzngard) [inline src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.vb:39]: nit: you're just digging into through `symbol.ElementType` here and above. Can we just combine into a single loop?

File: src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.vb [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            End Function
+
+            Private Shared Function CreateSimpleTypeSyntax(symbol As INamedTypeSymbol) As TypeSyntax
+                Dim syntax = TryCreateSpecializedNamedTypeSyntax(symbol)
+                If syntax IsNot Nothing Then
+                    Return syntax
+                End If
+
+                If symbol.IsTupleType AndAlso
+               symbol.TupleUnderlyingType IsNot Nothing AndAlso
```
Comment (ryzngard) [inline src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.vb:110]: nit: spacing looks weird here

File: src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.vb [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+                                                                                    GetTypeSyntaxForSymbol(element.Type))),
+                                                        DirectCast(SyntaxFactory.TypedTupleElement(
+                                                                        GetTypeSyntaxForSymbol(element.Type)), TupleElementSyntax)))))
+            End Function
+
+            Private Shared Function ToIdentifierName(text As String) As IdentifierNameSyntax
+                Return SyntaxFactory.IdentifierName(ToIdentifierToken(text)).WithAdditionalAnnotations(Simplifier.Annotation)
+            End Function
+
+            Private Shared Function ToIdentifierToken(text As String, Optional afterDot As Boolean = False, Optional symbol As ISymbol = Nothing, Optional withinAsyncMethod As Boolean = False) As SyntaxToken
```
Comment (ryzngard) [inline src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.vb:194]: Same comments about `SyntaxFacts` as csharp case

**Review by ryzngard — APPROVED** [review]
Overall looks good. Minor clarifications and questions about moving some code to `SyntaxFacts`. Surprised it took this much complicated visitor code to get the correct element type and construct an array...

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs (hunk 3 above)
Comment (jmarolf) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpForwardCancellationTokenToInvocationsFixer.TypeNameVisitor.cs:219]: we don't have a `SyntaxFacts` type that is useable in analyzers for this repo. not a bad thing to add in a subsequent PR

**Pull request merged**